plugin/evm/vm.go

```diff
@@ -1271,6 +1279,52 @@ func (vm *VM) initBlockBuilding() error {
 	return nil
 }
 
+func (vm *VM) SubscribeToEvents(ctx context.Context) commonEng.Message {
+	// Block building is not initialized yet, so we haven't finished syncing or bootstrapping.
+	if vm.builder == nil {
+		select {
+		case <-ctx.Done():
+			return 0
+		case ss := <-vm.stateSyncDone:
+			return ss
+		case <-vm.shutdownChan:
+			return 0
+		case <-vm.finishedSyncing:
+			break
+		}
+	}
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+	defer wg.Wait()
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer func() {
+		cancel()
+		vm.builder.wakeup()
+	}()
+
+	pending := make(chan commonEng.Message, 1)
+
+	go func() {
+		defer wg.Done()
+		pending <- vm.builder.waitForTxEnqueue(ctx)
+	}()
+
+	select {
+	case ss := <-vm.stateSyncDone:
+		return ss
+	case pendingTx := <-pending:
+		return pendingTx
+	case <-ctx.Done():
+		vm.builder.wakeup()
```

Comment:
why do we need to wakeup/broadcast the builder signal?

Reply:
The goroutine [here](https://github.com/yacovm/coreth/blob/04b8c147292353c3fb000252796ea9575f9e459c/plugin/evm/vm.go#L1311) needs to finish and if the caller gives up on the invocation by cancelling the context, or if the VM is shut down, then we must unblock the goroutine waiting on the condition variable.